azure-iot-device/tests/iothub/pipeline/test_iothub_pipeline.py

```diff
@@ -120,66 +130,89 @@ def test_pipeline_configuration(self, auth_provider):
         "Runs a SetAuthProviderOperation with the provided AuthenticationProvider on the pipeline, if using SAS based authentication"
     )
     def test_sas_auth(self, mocker, device_connection_string):
-        mocker.patch.object(pipeline_stages_base, "PipelineRootStage")
+        mocker.spy(pipeline_stages_base.PipelineRootStage, "run_op")
         auth_provider = SymmetricKeyAuthenticationProvider.parse(device_connection_string)
         pipeline = IoTHubPipeline(auth_provider)
-        op = pipeline._pipeline.run_op.call_args[0][0]
+        op = pipeline._pipeline.run_op.call_args[0][1]
         assert pipeline._pipeline.run_op.call_count == 1
         assert isinstance(op, pipeline_ops_iothub.SetAuthProviderOperation)
         assert op.auth_provider is auth_provider
 
     @pytest.mark.it(
         "Propagates exceptions that occurred in execution upon unsuccessful completion of the SetAuthProviderOperation"
     )
-    def test_sas_auth_op_fail(self, mocker, device_connection_string):
-        mocker.patch.object(pipeline_stages_base, "PipelineRootStage")
-        auth_provider = SymmetricKeyAuthenticationProvider.parse(device_connection_string)
-        pipeline = IoTHubPipeline(auth_provider)
-        op = pipeline._pipeline.run_op.call_args[0][0]
-        op.error = Exception()
+    def test_sas_auth_op_fail(self, mocker, device_connection_string, fake_exception):
+        old_execute_op = pipeline_stages_base.PipelineRootStage._execute_op
```

Comment:
Not gained accuracy, but a different requirement on the behavior of the pipeline.  The IoTHubPipeline initializer can't complete until PipelineRootStage.run_op() calls it's callback.  In the old code, this wasn't necessary.
This has to do with your comment about mocking the transport  -- i need to go back and refactor this to mock the pipeline (which was done in the past) instead of the transport (which is done now).

---
In reply to: [318848210](https://github.com/Azure/azure-iot-sdk-python-preview/pull/174#discussion_r318848210) [](ancestors = 318848210)